pkg/controller/kubedirectorcluster/cluster_restore.go

```diff
@@ -0,0 +1,411 @@
+// status exists.
+func clusterServiceExists(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+) bool {
+
+	clusterService := cr.Status.ClusterService
+	if clusterService != "" {
+		_, serviceErr := observer.GetService(
+			cr.Namespace,
+			clusterService,
+		)
+		if serviceErr != nil {
+			shared.LogInfof(
+				reqLogger,
+				cr,
+				shared.EventReasonCluster,
+				"being restored: cluster service %s does not exist",
+				clusterService,
+			)
+			return false
+		}
+	}
+	return true
+}
+
+// roleResourcesExist looks to see if a statefulsets named in the
+// status exist, along with the necessary per-member services.
+func roleResourcesExist(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+) bool {
+
+	for _, roleStatus := range cr.Status.Roles {
+		if roleStatus.StatefulSet != "" {
+			_, statefulSetErr := observer.GetStatefulSet(
+				cr.Namespace,
+				roleStatus.StatefulSet,
+			)
+			if statefulSetErr != nil {
+				shared.LogInfof(
+					reqLogger,
+					cr,
+					shared.EventReasonCluster,
+					"being restored: statefulset %s does not exist",
+					roleStatus.StatefulSet,
+				)
+				return false
+			}
+		}
+		for _, memberStatus := range roleStatus.Members {
+			memberService := memberStatus.Service
+			if memberService != "" && memberService != zeroPortsService {
+				_, serviceErr := observer.GetService(
+					cr.Namespace,
+					memberService,
+				)
+				if serviceErr != nil {
+					shared.LogInfof(
+						reqLogger,
+						cr,
+						shared.EventReasonCluster,
+						"being restored: member service %s does not exist",
+						memberService,
+					)
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
+// checkRestoreDone will try to clear the being-restored label if all the
+// "waiting" flags are false. If that fails it will populate the error field
+// of the restore progress object. Return true if and only if we have
+// successfully cleared that label.
+func checkRestoreDone(
+	reqLogger logr.Logger,
+	cr *kdv1.KubeDirectorCluster,
+) bool {
+
+	if cr.Status.RestoreProgress.AwaitingApp ||
+		cr.Status.RestoreProgress.AwaitingStatus ||
+		cr.Status.RestoreProgress.AwaitingResources {
+		return false
+	}
+
+	// Let's not deep-copy the whole CR; we just need to modify Labels.
+	patchedCR := *cr
+	patchedCR.Labels = make(map[string]string)
+	for key, value := range cr.Labels {
+		if key != shared.RestoringLabel {
+			patchedCR.Labels[key] = value
+		}
+	}
+	patchErr := shared.Patch(
+		context.TODO(),
+		cr,
+		&patchedCR,
+	)
+	if patchErr == nil {
+		shared.LogInfo(
+			reqLogger,
+			cr,
+			shared.EventReasonCluster,
+			"resuming reconciliation",
+		)
+		return true
+	}
+	// Don't want to use LogError here because it generates a stacktrace.
+	// We don't really care about that, just about the message.
+	shared.LogInfof(
+		reqLogger,
+		cr,
+		shared.EventReasonCluster,
+		"failed to resume reconciliation: %s",
+		patchErr.Error(),
+	)
+	cr.Status.RestoreProgress.Error = patchErr.Error()
```

Comment:
shouldn't all the flags be marked false here?

Reply:
The flags' true-ness still represents the known state of things, we just failed to automatically resume reconciliation.